### PR TITLE
Use hir::ItemLocalId as keys in TypeckTables.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR makes `TypeckTables` use `ItemLocalId` instead of `NodeId` as key. This is needed for incremental compilation -- for stable hashing and for being able to persist and reload these tables. The PR implements the most important part of https://github.com/rust-lang/rust/issues/40303. 

Some notes on the implementation:
* The PR adds the `HirId` to HIR nodes where needed (`Expr`, `Local`, `Block`, `Pat`) which obviates the need to store a `NodeId -> HirId` mapping in crate metadata. Thanks @eddyb for the suggestion! In the future the `HirId` should completely replace the `NodeId` in HIR nodes.
* Before something is read or stored in one of the various `TypeckTables` subtables, the entry's key is validated via the new `TypeckTables::validate_hir_id()` method. This makes sure that we are not mixing information from different items in a single table.

That last part could be made a bit nicer by either (a) new-typing the table-key and making `validate_hir_id()` the only way to convert a `HirId` to the new-typed key, or (b) just encapsulate sub-table access a little better. This PR, however, contents itself with not making things significantly worse.

Also, there's quite a bit of switching around between `NodeId`, `HirId`, and `DefIndex`. These conversions are cheap except for `HirId -> NodeId`, so if the valued reviewer finds such an instance in a performance critical place, please let me know. 

Ideally we convert more and more code from `NodeId` to `HirId` in the future so that there are no more `NodeId`s after HIR lowering anywhere. Then the amount of switching should be minimal again.

r? @eddyb, maybe?

### CODE REVIEW

**Comment by eddyb** [comment]
LGTM. r? @arielb1

File: src/librustc/middle/dead.rs [hunk 1]
```diff
@@ -119,6 +119,8 @@ impl<'a, 'tcx> MarkSymbolVisitor<'a, 'tcx> {
 
     fn handle_field_pattern_match(&mut self, lhs: &hir::Pat, def: Def,
                                   pats: &[codemap::Spanned<hir::FieldPat>]) {
+
```
Comment (arielb1) [inline src/librustc/middle/dead.rs]: ?

File: src/librustc/middle/reachable.rs [hunk 2]
```diff
@@ -375,7 +375,7 @@ fn reachable_set<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, crate_num: CrateNum) ->
     });
     let mut reachable_context = ReachableContext {
         tcx,
-        tables: &ty::TypeckTables::empty(),
+        tables: &ty::TypeckTables::empty(DefId::invalid()),
```
Comment (arielb1) [inline src/librustc/middle/reachable.rs]: I think this should be a FIXME? Shouldn't `tables` be `None` when outside of an item?

File: src/librustc/ty/context.rs [hunk 3]
```diff
@@ -272,9 +275,10 @@ pub struct TypeckTables<'tcx> {
 }
 
 impl<'tcx> TypeckTables<'tcx> {
-    pub fn empty() -> TypeckTables<'tcx> {
+    pub fn empty(local_id_root: DefId) -> TypeckTables<'tcx> {
         TypeckTables {
-            type_dependent_defs: NodeMap(),
+            local_id_root,
```
Comment (arielb1) [inline src/librustc/ty/context.rs]: The thing with `local_id_root` being sometimes invalid makes me uncomfortable. Please insert an `Option` *somewhere* (I think making `local_id_root` an option would work well).

File: src/librustc/ty/context.rs [hunk 4]
```diff
@@ -384,6 +390,30 @@ impl<'tcx> TypeckTables<'tcx> {
     pub fn upvar_capture(&self, upvar_id: ty::UpvarId) -> ty::UpvarCapture<'tcx> {
         self.upvar_capture_map[&upvar_id]
     }
+
+    /// Validate that a NodeId can safely be converted to an ItemLocalId for
```
Comment (arielb1) [inline src/librustc/ty/context.rs]: When is this function called? It seems to be called on some random HIR id accesses.

File: src/librustc/hir/map/definitions.rs [hunk 5]
```diff
@@ -466,7 +466,11 @@ impl Definitions {
     }
 
     pub fn find_node_for_hir_id(&self, hir_id: hir::HirId) -> ast::NodeId {
```
Comment (arielb1) [inline src/librustc/hir/map/definitions.rs]: Can't you rather do some sort of merging of this function with `validate_hir_id`? Unmarked linear searches are dangerous.

File: src/librustc/ty/context.rs [hunk 6]
```diff
@@ -207,52 +207,55 @@ pub struct CommonTypes<'tcx> {
 
 #[derive(RustcEncodable, RustcDecodable)]
 pub struct TypeckTables<'tcx> {
+    /// The HirId::owner all ItemLocalIds in this table are relative to.
+    pub local_id_root: DefId,
```
Comment (arielb1) [inline src/librustc/ty/context.rs]: This is in several places initialized with an `invalid_def_id`. Why can't this be an `Option`?

**Comment by arielb1** [comment]
I don't like the way `validate_def_id` is spread around randomly in code. I think I would make the `TypeckTables` fields to be functions that return a
```Rust
pub struct LocalTableInContext<'a, V> {
    parent_def_id: Option<DefId>,
    data: &'a LocalItemMap<V>
}
```

Which does validation on the accessor methods. Or something.

**Comment by michaelwoerister** [comment]
OK, I'll change the various visitors to use an `Option<&TypeckTables>`. I don't like make the `local_id_root` optional, since every `TypeckTables` should really have one of these.

I'll also look into a way of handling `HirId` validation in a more central place.

**Comment by michaelwoerister** [comment]
@arielb1 I think I addressed your comments.

**Comment by michaelwoerister** [comment]
(Note: I've added the last three commits. The others were not touched except for rebasing).

File: src/librustc/hir/def_id.rs [hunk 7]
```diff
@@ -201,4 +201,11 @@ impl DefId {
     pub fn is_local(&self) -> bool {
         self.krate == LOCAL_CRATE
     }
+
```
Comment (arielb1) [inline src/librustc/hir/def_id.rs]: unused function plz kill

File: src/librustc/middle/dead.rs [hunk 8]
```diff
@@ -119,7 +119,9 @@ impl<'a, 'tcx> MarkSymbolVisitor<'a, 'tcx> {
 
     fn handle_field_pattern_match(&mut self, lhs: &hir::Pat, def: Def,
                                   pats: &[codemap::Spanned<hir::FieldPat>]) {
-        let variant = match self.tables.node_id_to_type(lhs.id).sty {
+
```
Comment (arielb1) [inline src/librustc/middle/dead.rs]: nit: stray newlines in program

File: src/librustc/ty/context.rs [hunk 9]
```diff
@@ -207,54 +211,153 @@ pub struct CommonTypes<'tcx> {
...
+    }
+}
+
+pub struct LocalTableInContextMut<'a, V: 'a> {
+    local_id_root: Option<DefId>,
+    data: &'a mut ItemLocalMap<V>
+}
+
+impl<'a, V> LocalTableInContextMut<'a, V> {
+
```
Comment (arielb1) [inline src/librustc/ty/context.rs]: nit: stray newline in program

File: src/librustc_privacy/lib.rs [hunk 10]
```diff
@@ -1606,21 +1656,26 @@ fn privacy_access_levels<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
     let krate = tcx.hir.krate();
 
+    let empty_tables = ty::TypeckTables::empty(None);
+
```
Comment (arielb1) [inline src/librustc_privacy/lib.rs]: nit: stray newline in program (I think)

**Comment by arielb1** [comment]
Nice! r=me modulo nits.

File: src/librustc/ty/context.rs (hunk 9 above)
Comment (michaelwoerister) [inline src/librustc/ty/context.rs]: I actually like newlines in places like this. Makes the `impl` line easier to read. But I'll remove it for consistency.

**Comment by michaelwoerister** [comment]
@bors r=arielb1

Amended the nit fixes to the last commit. Thanks for the reviews!

File: src/librustc/ty/context.rs [hunk 11]
```diff
@@ -207,54 +211,152 @@ pub struct CommonTypes<'tcx> {
...
+/// Validate that the given HirId (respectively its `local_id` part) can be
+/// safely used as a key in the tables of a TypeckTable. For that to be
+/// the case, the HirId must have the same `owner` as all the other IDs in
+/// this table (signified by `local_id_root`). Otherwise the HirId
+/// would be in a different frame of reference and using its `local_id`
+/// would result in lookup errors, or worse, in silently wrong data being
+/// stored/returned.
+fn validate_hir_id_for_typeck_tables(local_id_root: Option<DefId>,
+                                     hir_id: hir::HirId,
+                                     mut_access: bool) {
```
Comment (kennytm) [inline src/librustc/ty/context.rs]:
```
[00:10:17] error: unused variable: `local_id_root`
[00:10:17]    --> /checkout/src/librustc/ty/context.rs:226:38
[00:10:17]     |
[00:10:17] 226 | fn validate_hir_id_for_typeck_tables(local_id_root: Option<DefId>,
[00:10:17]     |                                      ^^^^^^^^^^^^^
[00:10:17]     |
[00:10:17] 
[00:10:17] error: unused variable: `hir_id`
[00:10:17]    --> /checkout/src/librustc/ty/context.rs:227:38
[00:10:17]     |
[00:10:17] 227 |                                      hir_id: hir::HirId,
[00:10:17]     |                                      ^^^^^^
[00:10:17] 
[00:10:17] error: unused variable: `mut_access`
[00:10:17]    --> /checkout/src/librustc/ty/context.rs:228:38
[00:10:17]     |
[00:10:17] 228 |                                      mut_access: bool) {
[00:10:17]     |                                      ^^^^^^^^^^
[00:10:17] 
[00:10:18] error: aborting due to 3 previous errors
[00:10:18] 
[00:10:18] error: Could not compile `rustc`.
```

Comment (michaelwoerister) [inline src/librustc/ty/context.rs]: Thanks, @kennytm! That must be because of the `cfg(debug_assertions)` switch...

**Pull request merged**